### PR TITLE
feat: TE7-T122 Active Record Associations

### PR DESCRIPTION
**Pull request by Krutik31**
Create Event CRUD app with name, description, event_date(date) Users can log in & log out and create their own event lists.(using belongs_to and has_many associations) Display all the Events on the root page(index page) in decreasing order User can enrol in many different event and event contains many users(using many_to_many association) Makes user's profile page(User can has one profile detail). On User's profile page display the events which is enrolled by the user. Add a user's address. Store user's address in different table called 'addresses'. This table must contain a foreign key which is by default named after the relation (make association like user has_one: address, :as => :addressable) Add a category model to add events by its category(Like, music, visual arts, Film, Fashion etc.) Add functionality to filter events by its category. Add a comment MVC in which user can comment on event(using polymorphic association) Using one comment model to comment on multiple events.(User can like other's user's comment also) On the event show page display all the comments.

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class CommentsController < ApplicationController
+  before_action :login_checkup
+
+  def index
+    @comments = Comment.where(event_id: params[:id])
+    @event = Event.find(params[:id])
+  end
+
+  def add_comment
+    @comment = Comment.new(event_id: params[:id], user_id: session[:current_user_id], comment: params[:comment])
```
Comment (purvak-simform) [inline app/controllers/comments_controller.rb]: Why not created strong params for it

File: app/controllers/enrollments_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+class EnrollmentsController < ApplicationController
+  before_action :login_checkup
+
+  def index
+    @enrolled_events = Enrollment.where(user_id: session[:current_user_id], created: false).pluck(:event_id)
+    @events = Event.find(Enrollment.where(created: true).where.not(user_id: session[:current_user_id]).pluck(:event_id))
```
Comment (purvak-simform) [inline app/controllers/enrollments_controller.rb]: Here you are finding all the events that current user has attended so you can query it like this

@events = current_user.events.where(created: true)

File: app/controllers/enrollments_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  def create
+    @enrollment = Enrollment.new(event_id: params[:eventid], user_id: session[:current_user_id],
```
Comment (purvak-simform) [inline app/controllers/enrollments_controller.rb]: Don't use attribute name similar to built in name change **created** to something else like **enrolled_date**

File: app/controllers/events_controller.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  def new
+    @event = Event.new
+  end
+
+  def create
+    @event = Event.new(event_params)
+
+    if @event.save
+      Enrollment.create(event_id: @event.id, user_id: session[:current_user_id], created: true)
```
Comment (purvak-simform) [inline app/controllers/events_controller.rb]: move it to callback

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    flash[:notice] = 'You have to first login.'
+    redirect_to users_path
+  end
+
+  def find_event
+    @event = Event.find_by(id: Enrollment.where(user_id: session[:current_user_id],
+                                                event_id: params[:id]).pluck(:event_id))
+  end
+
+  def find_category
```
Comment (purvak-simform) [inline app/controllers/events_controller.rb]: 1. You are fetching the list of categories not finding category, so keep the method name accordingly
2. Use association to get records

File: app/controllers/events_controller.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    @categories = Category.all
+  end
+
+  def all_events
+    @events = Event.where(id: Enrollment.where(user_id: session[:current_user_id],
```
Comment (purvak-simform) [inline app/controllers/events_controller.rb]: Find records using association

File: app/controllers/comments_controller.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def like_comment
+    like = Like.find_or_create_by(user_id: session[:current_user_id], comment_id: params[:id])
+    eventid = like.comment.event_id
+    redirect_to show_comments_path(eventid)
+  end
+
+  def unlike_comment
+    @like = Like.find_by(user_id: session[:current_user_id], comment_id: params[:id])
```
Comment (purvak-simform) [inline app/controllers/comments_controller.rb]: use association here to get the like of user
current_user.likes.find_by(comment_id: params[:id])

File: app/controllers/enrollments_controller.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+class EnrollmentsController < ApplicationController
+  before_action :login_checkup
+
+  def index
+    @enrolled_events = Enrollment.where(user_id: session[:current_user_id], created: false).pluck(:event_id)
```
Comment (purvak-simform) [inline app/controllers/enrollments_controller.rb]: same find using association

File: app/controllers/enrollments_controller.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    if @enrollment.save
+      flash[:notice] = 'You have successfully enrolled.'
+      redirect_to enrollments_path
+    else
+      render :index, status: :unprocessable_entity
+    end
+  end
+
+  def discard
+    @enrollment = Enrollment.find_by(event_id: params[:eventid], user_id: session[:current_user_id],
```
Comment (purvak-simform) [inline app/controllers/enrollments_controller.rb]: use association

File: app/views/comments/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+                  <% else %>
+                    <%= link_to "Like (#{comment.likes.count})", like_comments_path(comment.id) %>
+                  <% end %>
+              </td>
+            </tr>
+          <% end %>
+        </tbody>
+      </table>
+
+      <div class="modal fade" id="addcomment">
```
Comment (purvak-simform) [inline app/views/comments/index.html.erb]: create partial for modal

File: app/views/comments/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+          <% @comments.each do |comment| %>
+            <tr>
+              <td class="text-left">
+                  <strong>User:</strong> <%= comment.user.username.titleize %>
+                  <br>
+                  <strong>At:</strong> <%= comment.created_at %>
+                  <br>
+                  <strong>Comment:</strong> <%= comment.comment %>
+                  <br>
+                  <% if comment.likes.pluck(:user_id).include?(session[:current_user_id]) %>
```
Comment (purvak-simform) [inline app/views/comments/index.html.erb]: Don't fire the query on the view page, use a controller for it.
and you can change it to something like this
Like.find_by(comment_id: comment.id, user_id: session[:current_user_id]).present?

File: app/views/enrollments/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            <th>Name</th>
+            <th>Description</th>
+            <th>Category</th>
+            <th>Event Date</th>
+            <th>Actions</th>
+          </tr>
+        </thead>
+        <tbody>
+          <% n = 1 %>
+          <% @events.each do |event| %>
```
Comment (purvak-simform) [inline app/views/enrollments/index.html.erb:26]: why used n = 1, what is requirement of it?

**Review by purvak-simform — CHANGES_REQUESTED** [review]

File: app/views/enrollments/index.html.erb (hunk 12 above)
Comment (Krutik31) [reply app/views/enrollments/index.html.erb:26]: It used to give an index numbers to the entries.

File: app/controllers/comments_controller.rb [hunk 13]
```diff
@@ -35,4 +35,8 @@ def login_checkup
     flash[:notice] = 'You have to first login.'
     redirect_to users_path
   end
+
+  def comment_params
+    params.require(:comment).permit(:commenttext)
```
Comment (purvak-simform) [inline app/controllers/comments_controller.rb:40]: Always keep attributes name in snake_case
**comment_text**

File: app/controllers/comments_controller.rb [hunk 14]
```diff
@@ -2,26 +2,26 @@ class CommentsController < ApplicationController
...
 
     return unless @comment.save
 
     redirect_to show_comments_path(params[:id])
   end
 
   def like_comment
-    like = Like.find_or_create_by(user_id: session[:current_user_id], comment_id: params[:id])
+    like = current_user.likes.find_or_create_by(comment_id: params[:id])
     eventid = like.comment.event_id
```
Comment (purvak-simform) [inline app/controllers/comments_controller.rb:19]: Variable name also be snake_case

File: app/views/comments/_modal.html.erb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    
+      <!-- Modal Header -->
+      <div class="modal-header">
+        <h4 class="modal-title"><%= @event.name %></h4>
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+      </div>
+      
+      <!-- Modal body -->
+      <div class="modal-body">
+        <%= form_with url: add_comments_path(@event.id), method: :get do |form| %>
```
Comment (purvak-simform) [inline app/views/comments/_modal.html.erb:13]: here there is no need to pass @event.id, just pass @event

**Review by purvak-simform — COMMENTED** [review]
@Krutik31 Update changes in your next PR

**Pull request merged**